cliquet/resource.py

```diff
@@ -19,37 +22,210 @@
+
+    readonly_methods = ('GET',)
+
+    service_arguments = {
+        'description': 'Collection of {resource_name}',
+        'cors_origins': ('*',),
+        'error_handler': json_error_handler
+    }
+
+    default_arguments = {
+        'cors_headers': ('Backoff', 'Retry-After', 'Alert'),
+    }
+
+
+    default_collection_arguments = {}
+    collection_get_arguments = {
+        'cors_headers': (('Backoff', 'Retry-After', 'Alert') +
+                        ('Next-Page', 'Total-Records', 'Last-Modified'))
+    }
+
+    default_record_arguments = {}
+    record_get_arguments = {
+        'cors_headers': (('Backoff', 'Retry-After', 'Alert') +
+                        ('Last-Modified',))
+    }
+
+    def __init__(self, **kwargs):
+        self.update(**kwargs)
+        self.record_arguments = functools.partial(
+                                        self.get_arguments, 'record')
+        self.collection_arguments = functools.partial(
+                                        self.get_arguments, 'collection')
+
+    def update(self, **kwargs):
+        self.__dict__.update(**kwargs)
+
+    def get_arguments(self, typ_, resource, method):
+        """Returns the arguments for the given type, where `typ_` can be
+        either "collection" or "record".
+        """
+        args = self.default_arguments.copy()
+        default_arguments = getattr(self, 'default_%s_arguments' % typ_)
+        args.update(**default_arguments)
+
+        by_method = '%s_%s_arguments' % (typ_, method.lower())
+        method_args = getattr(self, by_method, {})
+        args.update(**method_args)
 
-    By default, the lower class name of the resource is used to build URLs.
+        if method.lower() in map(str.lower, self.validate_schema_for):
+            args['schema'] = resource.mapping
+        else:
+            # Simply validate that posted body is a mapping.
+            args['schema'] = colander.MappingSchema(unknown='preserve')
+
+        permission = self.get_view_permission(typ_, resource, method)
+        if permission is not None:
+            args['permission'] = permission
+
+        return args
+
+    def get_view(self, typ_, method):
+        """Returns the view location for the given type and method.
+
+        For collections, this will be "collection_{method|lower}
+        For records, this will be "{method|lower}.
+        """
+        if typ_ == 'record':
+            return method.lower()
+        else:
+            return '%s_%s' % (typ_, method.lower())
 
-    This decorator accepts the same parameters as the :rtd:`Cornice <cornice>`
-    :meth:`~cornice:cornice.resource.resource` decorator.
+    def get_name(self, resource):
+        """Returns the name of the resource.
+        """
+        if 'name' in self.__dict__:
+            name = self.__dict__['name']
+        elif hasattr(resource, 'name') and not callable(resource.name):
+            name = resource.name
+        else:
+            name = resource.__name__.lower()
```

Comment:
``name`` is now an attribute of ``Collection`` and can be changed at any time with ``resource.collection.name = 'XX'``
```